pkg/controller/core/clusterqueue_controller.go

```diff
@@ -45,38 +47,89 @@ import (
 	"sigs.k8s.io/kueue/pkg/workload"
 )
 
+const snapshotWorkers = 5
+
 type ClusterQueueUpdateWatcher interface {
 	NotifyClusterQueueUpdate(*kueue.ClusterQueue, *kueue.ClusterQueue)
 }
 
 // ClusterQueueReconciler reconciles a ClusterQueue object
 type ClusterQueueReconciler struct {
-	client                client.Client
-	log                   logr.Logger
-	qManager              *queue.Manager
-	cache                 *cache.Cache
-	wlUpdateCh            chan event.GenericEvent
-	rfUpdateCh            chan event.GenericEvent
-	watchers              []ClusterQueueUpdateWatcher
-	reportResourceMetrics bool
+	client                               client.Client
+	log                                  logr.Logger
+	qManager                             *queue.Manager
+	cache                                *cache.Cache
+	snapshotsQueue                       workqueue.Interface
+	wlUpdateCh                           chan event.GenericEvent
+	rfUpdateCh                           chan event.GenericEvent
+	watchers                             []ClusterQueueUpdateWatcher
+	reportResourceMetrics                bool
+	queueVisibilityUpdateInterval        time.Duration
+	queueVisibilityClusterQueuesMaxCount int32
+}
+
+type ClusterQueueReconcilerOptions struct {
+	Watchers                             []ClusterQueueUpdateWatcher
+	ReportResourceMetrics                bool
+	QueueVisibilityUpdateInterval        time.Duration
+	QueueVisibilityClusterQueuesMaxCount int32
+}
+
+// Option configures the reconciler.
+type ClusterQueueReconcilerOption func(*ClusterQueueReconcilerOptions)
+
+func WithWatchers(watchers ...ClusterQueueUpdateWatcher) ClusterQueueReconcilerOption {
+	return func(o *ClusterQueueReconcilerOptions) {
+		o.Watchers = watchers
+	}
+}
+
+func WithReportResourceMetrics(report bool) ClusterQueueReconcilerOption {
+	return func(o *ClusterQueueReconcilerOptions) {
+		o.ReportResourceMetrics = report
+	}
+}
+
+// WithQueueVisibilityUpdateInterval specifies the time interval for updates to the structure
+// of the top pending workloads in the queues.
+func WithQueueVisibilityUpdateInterval(interval time.Duration) ClusterQueueReconcilerOption {
+	return func(o *ClusterQueueReconcilerOptions) {
+		o.QueueVisibilityUpdateInterval = interval
+	}
 }
 
+// WithQueueVisibilityClusterQueuesMaxCount indicates the maximal number of pending workloads exposed in the
+// cluster queue status
+func WithQueueVisibilityClusterQueuesMaxCount(value int32) ClusterQueueReconcilerOption {
+	return func(o *ClusterQueueReconcilerOptions) {
+		o.QueueVisibilityClusterQueuesMaxCount = value
+	}
+}
+
+var DefaultOptions = ClusterQueueReconcilerOptions{}
```

Comment:
Does it need to be exported?

Reply:
It does not need to be exported. But there is a collision with another `defaultOptions` in the package. We can rename it to `defaultCqOptions`.